---
title: "Lung CIS progressive vs regressive lesions WGS analysis"
author: "Henry Lee-Six"
date: "`r Sys.Date()`"
output: pdf_document
---

```{r global_options, include=FALSE}
knitr::opts_chunk$set(fig.width=8, fig.height=8, fig.path='Figs/',
                      echo=TRUE, include=TRUE, warning=FALSE, message=FALSE,
                      tidy.opts=list(width.cutoff=75),tidy=TRUE)
```

```{r set working directory, include=FALSE}
setwd("/Users/hl11/Documents/Projects/Lung_Janes_lab_collab/2017.12_resubmission/analysis_for_supplementary/")
```

```{r load libraries, include=T}
library("knitr")
library("lme4")
```

```{r Read in files with necessary information for modeling, include=FALSE}
modelinfo <- read.csv("/Users/hl11/Documents/Projects/Lung_Janes_lab_collab/2017.12_resubmission/burden_sigs/Model_info_all_except_subclonal_sigs_2017.12.14.txt",
                      sep="\t", header=T, stringsAsFactors = F)

```

```{r Show the files needed for modeling, include=T}
# This file has all the relevant information for modeling.
head(modelinfo)

```

```{r Write function for testing difference between lesions that went on to progress and those that went on to regress, include=TRUE}

compare.fn <- function(modelinfo, dependent_variable, compared_variable, fixed_effects, random_effects, offset=NULL) {
  
  # scale the numeric predictors 
  # first need to work out which are the numeric ones.
  nums <- sapply(modelinfo, is.numeric)
  fixedtoscale <- intersect(fixed_effects, names(nums[which(nums)]))
  fixednottoscale <- fixed_effects[!fixed_effects %in% names(nums[which(nums)])]
  
  # now do the scaling.
  scaleddat <- as.data.frame(cbind(modelinfo[,c(dependent_variable, compared_variable,random_effects, fixednottoscale)], scale(modelinfo[,c(fixedtoscale)])))
  scaleddat <- scaleddat[complete.cases(scaleddat),]
  
  # if the dependent variable is not an integer, round it.
  scaleddat[,dependent_variable] <- round(scaleddat[,dependent_variable], digits=0)
  
  # fit the full model
  fullform <- paste0(dependent_variable,  " ~ ", compared_variable, " + ", paste(fixed_effects, collapse=" + "), " + (1|", paste(random_effects, collapse=") + (1|"), ")" )
  full.model <- glmer(fullform, data=scaleddat, family="poisson", offset=offset)
  
  print("Full model")
  print(full.model)
  cat("\n")
  
  # fit the model removing the variable that is being assessed.
  reducedform <- paste0(dependent_variable,  " ~ ", paste(fixed_effects, collapse=" + "), " + (1|", paste(random_effects, collapse=") + (1|"), ")" )
  reduced.model <- glmer(reducedform, data=scaleddat, family="poisson", offset=offset)
  print("Reduced model")
  print(reduced.model)
  cat("\n")
  
  # compare the models using anova
  compare_anova <- anova(full.model, reduced.model)
  print("Comparing models")
  print(compare_anova)
  cat("\n")
  
  # plot the differences between the two groups
  # only plot if there is no offset. 
  if (is.null(offset)){
    par(mfrow=c(1,1))
    plotform <- paste0(dependent_variable, " ~ ", compared_variable)
    
    boxplot(modelinfo[,(colnames(modelinfo)==dependent_variable)] ~ modelinfo[,(colnames(modelinfo)==compared_variable)],
            add=F, notch=T, at=c(1.15,1.85), range=0, main=plotform, boxwex=1*1*0.5)
    stripchart(modelinfo[,(colnames(modelinfo)==dependent_variable)] ~ modelinfo[,(colnames(modelinfo)==compared_variable)],
               vertical=T, pch=16, col=c("black", "red"), ylab=dependant_variable,
               method="jitter", at=c(1.15,1.85), add=T, cex=2)
    legend("top", legend=c(paste0('LRT p.val = ', round(compare_anova$`Pr(>Chisq)`[2], digits=3))), bty="n")
  }
}
```

Now run the actual tests.

1) are there more substitutions in progressive than regressive cases?

```{r subs, include=TRUE}
compare.fn(dependent_variable = "subs", compared_variable = "outcome", fixed_effects = c("age", "packyears"), 
           random_effects = "patient", modelinfo=modelinfo)
```

1a) How does this change if just restrict to substitutions that were assigned to the clonal cluster in the Dirichlet process?

```{r clonal_subs, include=TRUE}
compare.fn(dependent_variable = "clonal_muts", compared_variable = "outcome", fixed_effects = c("age", "packyears"), 
           random_effects = "patient", modelinfo=modelinfo)
```

2) are there more indels in progressive than regressive cases?

```{r indels, include=TRUE}
compare.fn(dependent_variable = "indels", compared_variable = "outcome", fixed_effects = c("age", "packyears"), 
           random_effects = "patient", modelinfo=modelinfo)
```

3) are there more structural variants in progressive than regressive cases?

```{r SVs, include=T}
compare.fn(dependent_variable = "SVs", compared_variable = "outcome", fixed_effects = c("age", "packyears"), 
           random_effects = "patient", modelinfo=modelinfo)
```

4) are there more copy number changes in progressive than regressive cases?
```{r CN_changes, include=T}
compare.fn(dependent_variable = "copy.changes", compared_variable = "outcome", fixed_effects = c("age", "packyears"), 
           random_effects = "patient", modelinfo=modelinfo)
```

5) are there more drivers in progressive than regressive cases?
```{r drivers, include=T}
compare.fn(dependent_variable = "drivers", compared_variable = "outcome", fixed_effects = c("age", "packyears"), 
           random_effects = "patient", modelinfo=modelinfo)
```

5b) Does this hold true if control for the mutation rate?
For this only include drivers caused by substitutions or indels, and regress against the number of subs and indels

```{r drivers_no_cn, include=T}
compare.fn(dependent_variable = "drivers_no_CN", compared_variable = "outcome", fixed_effects = c("age", "packyears"), random_effects = "patient",
           offset=log(modelinfo$subs_plus_indels), modelinfo=modelinfo)
```

6) are there any differences in telomere lengths between progressive and regressive cases?

```{r telomeres, include=T}
compare.fn(dependent_variable = "telomeres", compared_variable = "outcome", fixed_effects = c("age", "packyears"), 
           random_effects = "patient", modelinfo=modelinfo)
```

7) are there any differences in the subclonal structure (either the number of subclones, or the proportion of subclonal mutations)?
NB for this have changed the fixed effects - I'm more concerned about whether coverage or cellularity could affect this than the age of the patient or their smoking history

```{r clone_number, include=T}
compare.fn(dependent_variable = "clone_number", compared_variable = "outcome", fixed_effects = c("cellularity", "sequencing_coverage"), 
           random_effects = "patient", modelinfo=modelinfo)
```

8) model the proportion of total subs that subclonal

```{r clonal_prop, include=T}
compare.fn(dependent_variable = "clonal_muts", compared_variable = "outcome", fixed_effects = c("cellularity", "sequencing_coverage"), random_effects = "patient",
           offset=log(modelinfo$subs), modelinfo=modelinfo)
```

9) are there any differences in the relative mutational signature contributions?

For each signature, compare, for progressive and regressive lesions:
a) the total number of mutations attributed to the signature
b) the proportion of mutations attributed to the signature (i.e. the counts but use an offset)

9.i.a) signature 1 counts.

```{r sig1counts, include=T}
compare.fn(dependent_variable = "Signature.1_counts", compared_variable = "outcome", fixed_effects = c("age", "packyears"), 
           random_effects = "patient", modelinfo=modelinfo)
```

9.i.b) signature 1 proportions.

```{r sig1props, include=T}
compare.fn(dependent_variable = "Signature.1_counts", compared_variable = "outcome", fixed_effects = c("age", "packyears"), 
           random_effects = "patient", modelinfo=modelinfo,
           offset=log(modelinfo$subs))
```

9.ii.a) signature 2 counts.

```{r sig2counts, include=T}
compare.fn(dependent_variable = "Signature.2_counts", compared_variable = "outcome", fixed_effects = c("age", "packyears"), 
           random_effects = "patient", modelinfo=modelinfo)
```

9.ii.b) signature 2 proportions.

```{r sig2props, include=T}
compare.fn(dependent_variable = "Signature.2_counts", compared_variable = "outcome", fixed_effects = c("age", "packyears"), 
           random_effects = "patient", modelinfo=modelinfo,
           offset=log(modelinfo$subs))
```

9.iii.a) signature 4 counts.

```{r sig4counts, include=T}
compare.fn(dependent_variable = "Signature.4_counts", compared_variable = "outcome", fixed_effects = c("age", "packyears"), 
           random_effects = "patient", modelinfo=modelinfo)
```

9.iii.b) signature 4 proportions.

```{r sig4props, include=T}
compare.fn(dependent_variable = "Signature.4_counts", compared_variable = "outcome", fixed_effects = c("age", "packyears"), 
           random_effects = "patient", modelinfo=modelinfo,
           offset=log(modelinfo$subs))
```

9.iv.a) signature 5 counts.

```{r sig5counts, include=T}
compare.fn(dependent_variable = "Signature.5_counts", compared_variable = "outcome", fixed_effects = c("age", "packyears"), 
           random_effects = "patient", modelinfo=modelinfo)
```

9.iv.b) signature 5 proportions.

```{r sig5props, include=T}
compare.fn(dependent_variable = "Signature.5_counts", compared_variable = "outcome", fixed_effects = c("age", "packyears"), 
           random_effects = "patient", modelinfo=modelinfo,
           offset=log(modelinfo$subs))
```

9.v.a) signature 13 counts.

```{r sig13counts, include=T}
compare.fn(dependent_variable = "Signature.13_counts", compared_variable = "outcome", fixed_effects = c("age", "packyears"), 
           random_effects = "patient", modelinfo=modelinfo)
```

9.v.b) signature 13 proportions.

```{r sig13props, include=T}
compare.fn(dependent_variable = "Signature.13_counts", compared_variable = "outcome", fixed_effects = c("age", "packyears"), 
           random_effects = "patient", modelinfo=modelinfo,
           offset=log(modelinfo$subs))
```